nano/node/wallet.cpp

```diff
@@ -1138,24 +1116,36 @@ std::shared_ptr<nano::block> nano::wallet::send_action (nano::account const & so
 
 	if (!error && block != nullptr && !cached_block)
 	{
-		if (nano::work_validate (*block))
+		if (action_complete (block, source_a, generate_work_a))
 		{
-			wallets.node.logger.try_log (boost::str (boost::format ("Cached or provided work for block %1% account %2% is invalid, regenerating") % block->hash ().to_string () % account_a.to_account ()));
-			wallets.node.work_generate_blocking (*block, wallets.node.active.limited_active_difficulty ());
+			// Return null block after work generation or ledger process error
+			block = nullptr;
 		}
-		wallets.watcher->add (block);
-		error = (wallets.node.process_local (block).code != nano::process_result::progress);
-		if (!error && generate_work_a)
+	}
+	return block;
+}
+
+bool nano::wallet::action_complete (std::shared_ptr<nano::block> block_a, nano::account const & account_a, bool const generate_work_a)
```

Comment:
`block_a` can be passed by const ref &

Reply:
Need to pass as shared_ptr to  `wallets.watcher->add` and `wallets.node.process_local`

Reply:
That's fine, those functions will copy (and extend lifetime) of the `shared_ptr` if necessary, but you don't need to copy it here.